### PR TITLE
Move listFeature to iterableLike

### PR DESCRIPTION
**Pull request by constantine-zubrylin**
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/main/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

**Review by robstoll — COMMENTED** [review]
did not take a deep look as it is a draft. let me know in case I shall already review

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/iterableFeatureExtractors.kt [hunk 1]
```diff
@@ -59,3 +61,29 @@ fun <E : Comparable<E>, T : Iterable<E>> Expect<T>.max(): Expect<E> =
...
+/**
+ * Expects that Iterable is not empty and returns an [Expect] for the last element in it.
+ *
+ * @return The newly created [Expect] for last element.
+ *
+ * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.lastFeature
+ *
+ * @since 1.2.0
+ */
+val <E, T: Iterable<E>> Expect<T>.last: Expect<E?>
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/iterableFeatureExtractors.kt:74]: no idea why you started adding `?` here and there but the implementation you had in List was just fine :)
```suggestion
val <E, T: Iterable<E>> Expect<T>.last: Expect<E>
```

File: logic/atrium-logic/src/generated/commonMain/ch/tutteli/atrium/logic/iterableLike.kt [hunk 2]
```diff
@@ -45,6 +45,9 @@ fun <T : IterableLike, E : Comparable<E>> AssertionContainer<T>.max(converter: (
 
 fun <T : IterableLike, E> AssertionContainer<T>.containsNoDuplicates(converter: (T) -> Iterable<E>): Assertion = impl.containsNoDuplicates(this, converter)
 
+fun <T : IterableLike, E : Comparable<E>> AssertionContainer<T>.last(converter: (T) -> Iterable<E>): FeatureExtractorBuilder.ExecutionStep<T, E> = impl.last(this, converter)
```
Comment (constantine-zubrylin) [inline logic/atrium-logic/src/generated/commonMain/ch/tutteli/atrium/logic/iterableLike.kt:48]: I could not manage to get rid of `E : Comparable<E>`. Need help with it.

Comment (robstoll) [reply logic/atrium-logic/src/generated/commonMain/ch/tutteli/atrium/logic/iterableLike.kt:48]: hm. `last` should not require comparable, what error do you get when you just use `E`

Comment (constantine-zubrylin) [reply logic/atrium-logic/src/generated/commonMain/ch/tutteli/atrium/logic/iterableLike.kt]: Screenshot 2024-03-10 at 23 21 33
Not sure if it's the best format to show an error :)

I've pushed a commit so you can have a look.

File: logic/atrium-logic/src/commonMain/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt [hunk 3]
```diff
@@ -250,4 +249,31 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
...
+    ): FeatureExtractorBuilder.ExecutionStep<T, E> =
+        container.extractFeature
+            .methodCall(method)
+            .withRepresentationForFailure(NO_ELEMENTS)
+            .withFeatureExtraction {
+                val iterable = converter(it)
+                Option.someIf(iterable.iterator().hasNext()) {
+                    iterable.collect() ?: throw IllegalStateException(
+                        "Iterable does not hasNext() even though checked before! Concurrent access?"
+                    )
```
Comment (robstoll) [inline logic/atrium-logic/src/commonMain/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt:273]: this is the problem, the last element could also be `null` we don't restrict it. Hence you need to change this to:
```suggestion
                    iterable.collect()
```

Comment (constantine-zubrylin) [reply logic/atrium-logic/src/commonMain/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt]: Oh, such a silly mistake. Thank you! I've pushed all the changes.

**Comment by robstoll** [comment]
@constantine-zubrylin thanks :+1:

**Pull request merged**